common/rpc/rpc.go

```diff
@@ -229,6 +232,14 @@ func (d *RPCFactory) CreateLocalFrontendGRPCConnection() *grpc.ClientConn {
 
 // CreateInternodeGRPCConnection creates connection for gRPC calls
 func (d *RPCFactory) CreateInternodeGRPCConnection(hostName string) *grpc.ClientConn {
+	connection := d.interNodeGrpcConnections.Get(hostName)
```

Comment:
Using a cached connection is a bit risky. What happens when the server closes the connection (restarts, scale up/down etc)? Won't the client in this case would try to use a closed connection?

You probably want to implement a full connection pool which understands the state of the connection and automatically manages re-establishing if needed.

Reply:
grpc manages all that stuff internally. a grpc.ClientConn is not 1:1 with a tcp connection.

and we never close these anyway. maybe we should close them, but when we do that we can just manually remove them from the cache.

Reply:
I see. Maybe this was the cause of https://github.com/temporalio/temporal/issues/6323. We were probably creating a lot of grpc client connection _channels_ and not closing them.
I'm curious to see if this caching will actually fix it (I think it will).